addons/mail/controllers/attachment.py

```diff
@@ -55,27 +55,23 @@ def mail_attachment_upload(self, ufile, thread_id, thread_model, is_pending=Fals
             raise NotFound()
         if thread_model == "discuss.channel" and not thread.allow_public_upload and not request.env.user._is_internal():
             raise AccessError(_("You are not allowed to upload attachments on this channel."))
-        vals = {
-            "name": ufile.filename,
-            "raw": ufile.read(),
-            "res_id": int(thread_id),
-            "res_model": thread_model,
-        }
         if is_pending and is_pending != "false":
             # Add this point, the message related to the uploaded file does
             # not exist yet, so we use those placeholder values instead.
-            vals.update(
-                {
-                    "res_id": 0,
-                    "res_model": "mail.compose.message",
-                }
-            )
+            vals = {"res_id": 0, "res_model": "mail.compose.message"}
+        else:
+            vals = {"res_id": int(thread_id), "res_model": thread_model}
         if request.env.user.share:
             # Only generate the access token if absolutely necessary (= not for internal user).
             vals["access_token"] = request.env["ir.attachment"]._generate_access_token()
         try:
             # sudo: ir.attachment - posting a new attachment on an accessible thread
-            attachment = request.env["ir.attachment"].sudo().create(vals)
+            attachment = request.env["ir.attachment"].sudo()._from_request_file(
+                ufile,
+                mimetype='TRUST' if request.env.user._is_internal() else 'GUESS',
+                res_id=vals.get("res_id"),
+                res_model=vals.get("res_model"),
```

Comment:
This needs to spread vals instead. No need to repeat the keys, and `access_token` will be missing.